### Chapter 14: ANOVA (Analysis of Variance)

In [2]:
#Two way ANOVA 

import pandas as pd
from statsmodels.formula.api import ols
from statsmodels.stats.anova import anova_lm
# Example data for two-way ANOVA

df= pd.DataFrame({
    'score': [
        70,72,75,78,80,82,
        65,68,70,71,73,75],
    'method': [
        'Traditional','Traditional','Online',
        'Online','Blended','Blended',
        'Traditional','Traditional','Online',
        'Online','Blended','Blended'],
    'class_size': [
        'Small','Small','Small',
        'Small','Small','Small',
        'Large','Large','Large',
        'Large','Large','Large'
    ]
})

In [3]:
import statsmodels.api as sm
from statsmodels.formula.api import ols

model = ols(
    'score ~ C(method) * C(class_size)',
    data=df
).fit()

anova_table = sm.stats.anova_lm(model, typ=2)

print(anova_table)

                             sum_sq   df          F    PR(>F)
C(method)                153.500000  2.0  29.709677  0.000771
C(class_size)            102.083333  1.0  39.516129  0.000754
C(method):C(class_size)    3.166667  2.0   0.612903  0.572525
Residual                  15.500000  6.0        NaN       NaN


Question 1 — One-way ANOVA

A company tests three sales training programs.


In [ ]:
# Determine:
# Whether the average sales performance differs between training programs.
# The ANOVA F-statistic.
# The p-value.
# If significant, perform Tukey HSD.
# Identify which training programs differ.


import pandas as pd
import statsmodels.api as sm
from statsmodels.formula.api import ols
from statsmodels.stats.multicomp import pairwise_tukeyhsd


df = pd.DataFrame({
    'score': [52, 55, 53, 54, 50, 56, 51, 60, 62, 61, 63, 59, 64, 62, 55, 57, 56, 58, 54, 59, 57],
    'program': ['A']*7 + ['B']*7 + ['C']*7
})

# Perform one-way ANOVA
model = ols('score ~ C(program)', data=df).fit()
anova_table = sm.stats.anova_lm(model, typ=2)
print(anova_table)

# If significant, perform Tukey HSD
tukey = pairwise_tukeyhsd(endog=df['score'], groups=df['program'], alpha=0.05)
print(tukey)

# Identify which training programs differ
print(tukey.summary())

                sum_sq    df          F        PR(>F)
C(program)  259.523810   2.0  36.824324  4.348285e-07
Residual     63.428571  18.0        NaN           NaN
Multiple Comparison of Means - Tukey HSD, FWER=0.05 
group1 group2 meandiff p-adj   lower   upper  reject
----------------------------------------------------
     A      B   8.5714    0.0  6.0106 11.1323   True
     A      C   3.5714  0.006  1.0106  6.1323   True
     B      C     -5.0 0.0003 -7.5608 -2.4392   True
----------------------------------------------------
Multiple Comparison of Means - Tukey HSD, FWER=0.05 
group1 group2 meandiff p-adj   lower   upper  reject
----------------------------------------------------
     A      B   8.5714    0.0  6.0106 11.1323   True
     A      C   3.5714  0.006  1.0106  6.1323   True
     B      C     -5.0 0.0003 -7.5608 -2.4392   True
----------------------------------------------------


Interpretation: 
Based on the ANOVA results, if the p-value is less than 0.05, we reject the null hypothesis and conclude that there are significant differences in average sales performance between the training programs. If the p-value is greater than 0.05, we fail to reject the null hypothesis and conclude that there is no significant difference in average sales performance between the training programs. 

If the ANOVA is significant, the Tukey HSD test helps identify which specific training programs differ from each other. The summary of the Tukey HSD test will show pairwise comparisons, the mean differences, and whether the differences are statistically significant. 

### Chi-Square Test: 
The Chi-Square test is used to determine whether there is a significant association between two categorical variables. It compares the observed frequencies in each category to the frequencies we would expect if there were no association between the variables.

In practice, you would create a contingency table of the two categorical variables and then apply the Chi-Square test to see if the observed distribution differs significantly from the expected distribution under the null hypothesis of no association.

In [5]:
# create contingency table for Chi-Square test 

contingency_table = pd.crosstab(df['program'], df['score']) 
print(contingency_table)

score    50  51  52  53  54  55  56  57  58  59  60  61  62  63  64
program                                                            
A         1   1   1   1   1   1   1   0   0   0   0   0   0   0   0
B         0   0   0   0   0   0   0   0   0   1   1   1   2   1   1
C         0   0   0   0   1   1   1   2   1   1   0   0   0   0   0


In [6]:
### Chi-Square Test Implementation
import pandas as pd
from scipy.stats import chi2_contingency

# perform Chi-Square test
chi2, p, dof, expected = chi2_contingency(contingency_table)
print(f"Chi2: {chi2}, p-value: {p}, Degrees of Freedom: {dof}")
print("Expected Frequencies:")
print(expected)

Chi2: 30.000000000000004, p-value: 0.3632178422794754, Degrees of Freedom: 28
Expected Frequencies:
[[0.33333333 0.33333333 0.33333333 0.33333333 0.66666667 0.66666667
  0.66666667 0.66666667 0.33333333 0.66666667 0.33333333 0.33333333
  0.66666667 0.33333333 0.33333333]
 [0.33333333 0.33333333 0.33333333 0.33333333 0.66666667 0.66666667
  0.66666667 0.66666667 0.33333333 0.66666667 0.33333333 0.33333333
  0.66666667 0.33333333 0.33333333]
 [0.33333333 0.33333333 0.33333333 0.33333333 0.66666667 0.66666667
  0.66666667 0.66666667 0.33333333 0.66666667 0.33333333 0.33333333
  0.66666667 0.33333333 0.33333333]]


In [7]:
from scipy.stats import chisquare

observed = [25, 35, 20, 20]
expected = [25, 25, 25, 25]

chi2, p = chisquare(
    f_obs=observed,
    f_exp=expected
)

print("Chi-square:", chi2)
print("p-value:", p)

Chi-square: 6.0
p-value: 0.11161022509471268


In [8]:
from scipy.stats import chi2_contingency

table = [
    [20, 30, 50],
    [35, 40, 25],
    [50, 35, 15]
]

chi2, p, dof, expected = chi2_contingency(table)

print("Chi-square:", chi2)
print("p-value:", p)
print("Degrees of freedom:", dof)
print("Expected frequencies:")
print(expected)

Chi-square: 35.952380952380956
p-value: 2.9597068766533027e-07
Degrees of freedom: 4
Expected frequencies:
[[35. 35. 30.]
 [35. 35. 30.]
 [35. 35. 30.]]


In [9]:
df = pd.DataFrame({
    "age_group": ["Young", "Senior", "Middle", "Young", "..."],
    "education": ["Bachelor", "HighSchool", "Master", "Bachelor", "..."],
    "employed": ["Yes", "Yes", "Yes", "No", "..."],
    "loan_approved": ["Yes", "No", "Yes", "No", "..."]
})

In [11]:
from scipy.stats import chi2_contingency 

table = pd.crosstab(df['education'], df['loan_approved'])
chi2, p, dof, expected = chi2_contingency(table)

print("Chi-square:", chi2)
print("p-value:", p)
print("Degrees of freedom:", dof)
print("Expected frequencies:")
print(expected) 

Chi-square: 7.5
p-value: 0.2770684433661073
Degrees of freedom: 6
Expected frequencies:
[[0.2 0.4 0.4]
 [0.4 0.8 0.8]
 [0.2 0.4 0.4]
 [0.2 0.4 0.4]]


### ANOVA Summary 20 September TA Session

Here is the **answer key with explanations**. The main skill in these questions is identifying **(1) the type of dependent variable, (2) whether groups are independent or related, and (3) whether you have expected proportions or two categorical variables.**

### QMEA Answer Key

| #      | Answer                                                                                                    | Why                                                                                                                    |
| ------ | --------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------- |
| **1**  | **C. One-Way ANOVA**                                                                                      | Compare the **mean** productivity score across **4 independent groups**.                                               |
| **2**  | **B. At least one population mean differs from another**                                                  | A significant ANOVA only tells us that **at least one mean is different**. It does not tell us which one.              |
| **3**  | **B. Randomized Block ANOVA**                                                                             | Same students measure all objects. **Objects = treatments; students = blocks.**                                        |
| **4**  | **C. To account for variation caused by a known nuisance factor**                                         | Blocking controls variation from a factor such as student, technician, patient, etc.                                   |
| **5**  | **C. The observations across treatments are related through the same participants**                       | Each participant tries all 3 diets → repeated/related observations → **blocks = participants**.                        |
| **6**  | **C. Chi-Square Goodness-of-Fit Test**                                                                    | One categorical variable is being compared with **specified expected proportions**.                                    |
| **7**  | **C. Chi-Square Test of Independence**                                                                    | Two categorical variables: **work shift × shirt condition**.                                                           |
| **8**  | **C. Determine whether customer choices among four brands follow proportions of 20%, 30%, 30%, and 20%.** | Goodness-of-fit tests **one categorical variable against expected proportions**.                                       |
| **9**  | **D. Chi-Square Test of Independence**                                                                    | Both variables are categorical: **department × satisfaction**.                                                         |
| **10** | **B. One-Way ANOVA**                                                                                      | 3 independent training groups with a numerical productivity score.                                                     |
| **11** | **B. Randomized Block ANOVA**                                                                             | Same 20 students use all 3 methods → **students are blocks**.                                                          |
| **12** | **B. Treatments = machines; Blocks = technicians**                                                        | We want to compare machines while controlling for technician differences.                                              |
| **13** | **A. Study A: One-Way ANOVA; Study B: Chi-Square Test of Independence**                                   | 0–100 score = quantitative → ANOVA. Satisfied/Not = categorical → chi-square independence.                             |
| **14** | **B. Study 1: Goodness-of-Fit; Study 2: Independence**                                                    | Study 1 compares observed vs expected proportions. Study 2 examines payment method × age group.                        |
| **15** | **C. First: One-Way ANOVA; Second: Randomized Block ANOVA**                                               | Different patients at each hospital → independent groups. Same patients under 3 systems → related measurements/blocks. |

## The key decision framework

This is probably the most useful way to memorize these for your QMEA exam:

### 1. One numerical outcome + independent groups → One-Way ANOVA

Ask:

> **Am I comparing means?**

Example:

> Productivity score of employees in Training A, B, C, D.

```text
Numerical outcome
       ↓
Compare means
       ↓
Different/independent people
       ↓
ONE-WAY ANOVA
```

**Example:**
50 patients at Hospital A, 50 different patients at Hospital B, 50 different patients at Hospital C.

→ **One-Way ANOVA**

---

### 2. Numerical outcome + same subjects/related groups → Randomized Block ANOVA

The key question:

> **Does the same person/object appear in multiple treatment groups?**

Example:

> 20 students try Method A, B, and C.

The students naturally differ in ability.

```text
Students
   ↓
Each student experiences
A + B + C
   ↓
Student = BLOCK
Method = TREATMENT
   ↓
Randomized Block ANOVA
```

This is essentially controlling for the variation caused by the **student**.

### Easy memory trick

**Same person repeatedly → BLOCK**

Examples:

* Same students → block
* Same patients → block
* Same technicians → block
* Same machines → potentially block, depending on the research question

---

### 3. One categorical variable + expected proportions → Goodness-of-Fit

Ask:

> **Do my observed categories match some expected percentages?**

Example:

A company claims:

```text
Package A = 10%
Package B = 20%
Package C = 30%
Package D = 20%
Package E = 20%
```

You collect actual customer choices.

You are testing:

$$
Observed \quad vs. \quad Expected
$$

→ **Chi-Square Goodness-of-Fit**

### Keyword

**Expected proportions → Goodness-of-Fit**

---

### 4. Two categorical variables → Chi-Square Independence

Ask:

> **Do I have two categorical variables and want to know whether they are associated?**

Example:

|          | Cash | Card | QR |
| -------- | ---: | ---: | -: |
| Under 30 |      |      |    |
| 30–49    |      |      |    |
| 50+      |      |      |    |

Variables:

* Age group = categorical
* Payment method = categorical

Question:

> Is payment method associated with age group?

→ **Chi-Square Test of Independence**

---

# The 4 tests in one picture

```text
                         START
                           │
                           ▼
                  What type of data?
                           │
             ┌─────────────┴─────────────┐
             │                           │
       Numerical outcome          Categorical outcome
             │                           │
             ▼                           ▼
       Comparing means?          How many categorical
             │                       variables?
       ┌─────┴─────┐             ┌──────┴──────┐
       │           │             │             │
  Independent   Same subjects   One           Two
    groups       / related      variable      variables
       │           │             │             │
       ▼           ▼             ▼             ▼
   One-Way      Randomized    Goodness-of-   Independence
    ANOVA         Block          Fit
                  ANOVA
```

## A very important distinction

### Goodness-of-Fit

**ONE categorical variable**

> Do customers choose A/B/C in the expected 20%/30%/50% proportions?

```text
Payment Method
      ↓
Cash / Card / QR
      ↓
Expected proportions?
      ↓
Goodness-of-Fit
```

### Independence

**TWO categorical variables**

> Is payment method associated with age group?

```text
Age Group + Payment Method
          ↓
      Association?
          ↓
    Independence Test
```

---

## ANOVA distinction

This is where Q1, Q3, Q5, Q10, Q11, Q12 and Q15 are testing you.

### One-Way ANOVA

Different people in each group:

```text
Group A       Group B       Group C
Person 1      Person 11     Person 21
Person 2      Person 12     Person 22
Person 3      Person 13     Person 23
   ↓             ↓             ↓
Independent observations
```

→ **One-Way ANOVA**

### Randomized Block ANOVA

Same people across treatments:

```text
             Treatment
             A    B    C
Student 1    ✓    ✓    ✓
Student 2    ✓    ✓    ✓
Student 3    ✓    ✓    ✓
Student 4    ✓    ✓    ✓

Student = BLOCK
A/B/C = TREATMENTS
```

→ **Randomized Block ANOVA**

---

# Q2 is particularly important

Suppose you have:

$$
H_0:\mu_1=\mu_2=\mu_3=\mu_4
$$

ANOVA gives:

$$
p < 0.05
$$

You reject \(H_0\).

The correct conclusion is:

> **At least one population mean differs from another.**

You **cannot** immediately say:

> "All four means are different."

You need a **post-hoc test**, such as **Tukey's HSD**, to determine which pairs differ.

For example:

```text
ANOVA significant
       ↓
At least one mean differs
       ↓
Tukey HSD
       ↓
Determine which pairs differ
```

This connects directly to the Tukey/ANOVA work you've been doing recently.

### Final answers to memorize

**1 C**
**2 B**
**3 B**
**4 C**
**5 C**
**6 C**
**7 C**
**8 C**
**9 D**
**10 B**
**11 B**
**12 B**
**13 A**
**14 B**
**15 C**

**Memory rule:**

> **Means + independent groups → One-Way ANOVA**
> **Means + same/related subjects → Randomized Block ANOVA**
> **One categorical variable + expected proportions → Goodness-of-Fit**
> **Two categorical variables + association → Chi-Square Independence**


In [2]:
# One -way ANOVA 

import pandas as pd
from scipy import stats

# Example data  
data = {
    'Finance': [3, 1, 4, 1],
    'Marketing': [1, 5, 3, 4],
    'Management': [8, 5, 4, 6]
}

df = pd.DataFrame(data)

# Perform one-way ANOVA 
# Ho: All group means are equal (null hypothesis)
# Ha: At least one group mean is different (alternative hypothesis)
f_statistic, p_value = stats.f_oneway(df['Finance'], df['Marketing'], df['Management'])
print("F-statistic:", f_statistic)
print("P-value:", p_value) 
print("Reject null hypothesis" if p_value < 0.05 else "Fail to reject null hypothesis")

F-statistic: 4.824742268041237
P-value: 0.037678234592938335
Reject null hypothesis


Conclusion  At least 5% significance level, the data provide sufficient evidence that the average number of job offers differs by MBA major. Management graduates received the highest sample average but ANOVA alone does not identify which specific groups differ from each other. Further post-hoc tests, such as Tukey's HSD, would be needed to determine the pairwise differences.


In [4]:
# post-hoc analysis 
from statsmodels.stats.multicomp import pairwise_tukeyhsd

# Perform Tukey's HSD test
tukey = pairwise_tukeyhsd(endog=df.melt()['value'], groups=df.melt()['variable'], alpha=0.05)
print(tukey)

# The summary table shows which pairs of groups have significantly different means. 

    Multiple Comparison of Means - Tukey HSD, FWER=0.05    
  group1     group2   meandiff p-adj   lower  upper  reject
-----------------------------------------------------------
   Finance Management      3.5 0.0353  0.2593 6.7407   True
   Finance  Marketing      1.0 0.6764 -2.2407 4.2407  False
Management  Marketing     -2.5 0.1335 -5.7407 0.7407  False
-----------------------------------------------------------


In [8]:
# indepedent two-sample t-test 

import scipy.stats as stats

t_statistic, p_value = stats.ttest_ind(df['Finance'], df['Management'])
print("T-statistic:", t_statistic)
print("P-value:", p_value)
print("Reject null hypothesis" if p_value < 0.05 else "Fail to reject null hypothesis")   # type: ignore

# which group has a higher average number of job offers  
# Hypothesis0: mean of Finance - mean of Management = 0.
# Hypothesis1: mean of Finance - mean of Management > 0.

# Perform one-tailed t-test for the hypothesis that Finance has a higher mean than Management
t_statistic, p_value = stats.ttest_ind(df['Finance'], df['Management'], alternative='greater')
print("One-tailed T-statistic:", t_statistic)
print("One-tailed P-value:", p_value)
print("Reject null hypothesis" if p_value < 0.05 else "Fail to reject null hypothesis")  #type: ignore

higher_avg_group = 'Finance' if df['Finance'].mean() > df['Management'].mean() else 'Management' 

print("Group with higher average number of job offers:", higher_avg_group)

T-statistic: -3.079589415448419
P-value: 0.02167440298102218
Reject null hypothesis
One-tailed T-statistic: -3.079589415448419
One-tailed P-value: 0.9891627985094888
Fail to reject null hypothesis
Group with higher average number of job offers: Management


### Exerise 3

In [10]:
# Randomized Block ANOVA: measurement error
import pandas as pd
from statsmodels.formula.api import ols
from statsmodels.stats.anova import anova_lm

measurements = pd.DataFrame({
    'Student': [1, 2, 3, 4],
    'Professor': [1.4, 3.1, 2.8, 3.4],
    'Male Student': [1.5, 2.6, 2.1, 3.6],
    'Female Student': [1.3, 2.4, 1.5, 2.9]
})

long_data = measurements.melt(
    id_vars='Student',
    var_name='Subject',
    value_name='Measurement_Error'
)

# Subject = treatment; Student = randomized block
model = ols(
    'Measurement_Error ~ C(Subject) + C(Student)',
    data=long_data
).fit()
anova_table = anova_lm(model, typ=2)

print(anova_table)
print(f"Treatment F-statistic: {anova_table.loc['C(Subject)', 'F']:.3f}")
print(f"Block F-statistic: {anova_table.loc['C(Student)', 'F']:.3f}")
print("Treatment decision:", "Reject H0" if anova_table.loc['C(Subject)', 'PR(>F)'] < 0.05 else "Do not reject H0") # type: ignore
print("Block decision:", "Reject H0" if anova_table.loc['C(Student)', 'PR(>F)'] < 0.05 else "Do not reject H0") # type: ignore

              sum_sq   df          F    PR(>F)
C(Subject)  0.871667  2.0   5.077670  0.051228
C(Student)  5.910000  3.0  22.951456  0.001092
Residual    0.515000  6.0        NaN       NaN
Treatment F-statistic: 5.078
Block F-statistic: 22.951
Treatment decision: Do not reject H0
Block decision: Reject H0


### Randomized Block ANOVA: Measurement Error

**a. Treatments and blocks**

- **Treatments:** The three subjects being measured: the professor, the male student, and the female student.
- **Blocks:** The four students who performed the measurements. Each student measured all three subjects, so student-to-student differences are controlled as blocks.

**b. Hypotheses**

Let $\mu_P$, $\mu_M$, and $\mu_F$ represent the mean measurement errors for the professor, male student, and female student.

- $H_0: \mu_P = \mu_M = \mu_F$
- $H_a:$ At least one mean measurement error differs.

**c. F-statistic**

From the ANOVA table,

$$
F = \frac{MS_{Treatments}}{MS_{Error}}
  = \frac{0.436}{0.086}
  \approx 5.078.
$$

**d. Decision at the 5% significance level**

The treatment p-value is $0.051$, which is greater than $\alpha=0.05$. Therefore, **do not reject $H_0$**. The block effect is significant because its p-value is $0.001$, but this does not change the treatment decision.

**e. Conclusion**

At the 5% significance level, there is **insufficient evidence to conclude that the mean measurement errors differ among the professor, male student, and female student**. The observed differences are not statistically significant at this level.